mlir/include/Catalyst/Driver/CompilerDriver.h

```diff
@@ -0,0 +1,115 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Support/LogicalResult.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/raw_ostream.h"
+#include <filesystem>
+#include <vector>
+
+/// Data about the JIT function that is optionally inferred and returned to the caller.
+///
+/// This is important for calling a function when invoking the compiler on an MLIR or LLVM textual
+/// representation intead of from Python.
+struct FunctionAttributes {
+    /// The name of the primary JIT entry point function.
+    std::string functionName;
+    /// The return type of the JIT entry point function.
+    std::string returnType;
+};
+
+/// Verbosity level
+// TODO: Adjust the number of levels according to our needs. MLIR seems to print few really
+// low-level messages, we might want to hide these.
+typedef enum {
+    CO_VERB_SILENT = 0,
+    CO_VERB_URGENT = 1,
+    CO_VERB_DEBUG = 2,
+    CO_VERB_ALL = 3
+} Verbosity;
+
+#define CO_MSG(opt, level, op)                                                                     \
+    {                                                                                              \
+        if ((opt).verbosity >= (level)) {                                                          \
+            options.diagnosticStream << op;                                                        \
+        }                                                                                          \
+    }
```

Comment:
Is there a need for this macro or could a function not do the same?

Reply:
I pass incomplete stream expressions as third argument. As far as I know, the idiomatic C++ would require us writing a function accepting a lambda, the usage would be like `MSG(options, CO_VERB_URGENT, [](stream s){ s << "the message"; })`. I think that the usual practice is to define macros instead, but let me know if you want another approach here.

Reply:
Hmm I see, I'll leave up to you then just wanted to see whether we could easily avoid the macro as we generally prefer a C++ implementation.

Reply:
Yes, I am not a big fan of macros, but I think the verbose printing is kind of exception becoming a common knowledge. Since MLIR also uses their version of the similar system, I would also leave them as-is. But I will be glad to switch to any pure-C++ solution when it became known.